### PR TITLE
cortex-m4_common: Refactor context save/restore

### PR DESCRIPTION
**Pull request by gebart**
Clang 3.6 is stricter than GCC and previous releases of Clang regarding
naked functions and now only allows inline asm inside naked functions (i.e. no C code).

This is a refactoring of the functions isr_pendsv, isr_svc,
context_save, context_restore to still be naked functions but allow
building with Clang 3.6

Building the current C/asm mixed naked function with Clang 3.6 results in:

    .../RIOT/cpu/cortex-m4_common/thread_arch.c:145:5: error: non-ASM statement in naked function is not supported
        enableIRQ();
        ^
    .../RIOT/core/include/arch/irq_arch.h:40:25: note: expanded from macro 'enableIRQ'
    #define enableIRQ       irq_arch_enable
                            ^
    .../RIOT/cpu/cortex-m4_common/thread_arch.c:142:16: note: attribute is here
    __attribute__((naked)) void NORETURN thread_arch_start_threading(void)
                   ^
    .../RIOT/cpu/cortex-m4_common/thread_arch.c:193:5: error: non-ASM statement in naked function is not supported
        sched_run();
        ^
    .../RIOT/cpu/cortex-m4_common/thread_arch.c:191:16: note: attribute is here
    __attribute__((naked)) void isr_svc(void)
                   ^
    .../RIOT/cpu/cortex-m4_common/thread_arch.c:202:5: error: non-ASM statement in naked function is not supported
        context_save();
        ^
    .../RIOT/cpu/cortex-m4_common/thread_arch.c:200:16: note: attribute is here
    __attribute__((naked)) void isr_pendsv(void)
                   ^
    2 warnings and 3 errors generated.

### CODE REVIEW

**Comment by haukepetersen** [comment]
Didn't test, but your code seems broken?! You are still calling the PendSV and the SVC interrupts, but you took away their implementations...

**Comment by gebart** [comment]
It's a weird construct, but I added the definitions of the isr_pendsv and isr_svc functions inside the inline asm block, since the pendsv had the same code as the svc, except saving the context before performing the scheduling.

File: cpu/cortex-m4_common/thread_arch.c [hunk 1]
```diff
@@ -156,50 +160,54 @@ void thread_arch_yield(void)
...
 /**
  * @brief The svc is used for running the scheduler and scheduling a new task during start-up or
  *        after a thread has exited
  */
-__attribute__((naked)) void isr_svc(void)
-{
-    sched_run();
-    context_restore();
-}
+void isr_svc(void);
```
Comment (haukepetersen) [inline cpu/cortex-m4_common/thread_arch.c]: this can be removed then?!

File: cpu/cortex-m4_common/thread_arch.c [hunk 2]
```diff
@@ -156,50 +160,54 @@ void thread_arch_yield(void)
...
 
 /**
- * @brief All task switching activity is carried out int the pendSV interrupt
+ * @brief All task switching activity is carried out in the PendSV interrupt
  */
-__attribute__((naked)) void isr_pendsv(void)
+void isr_pendsv(void);
```
Comment (haukepetersen) [inline cpu/cortex-m4_common/thread_arch.c]: so is this?!

**Comment by haukepetersen** [comment]
alright, I did not see that -> looks fine to me. The only thing I see, is that we might get in trouble, if we decide to use the SVC interrupt with different parameters for other purposes later on, but for one I think its not very probable, and for two let's worry about it then :-)

File: cpu/cortex-m4_common/thread_arch.c (hunk 1 above)
Comment (gebart) [inline cpu/cortex-m4_common/thread_arch.c]: I figured it can act as an anchor for the Doxygen description.

**Comment by haukepetersen** [comment]
Don't have an M4 board at hand right now, will test tomorrow.

**Comment by gebart** [comment]
Porting this to the cortex m3 should only be a matter of copy and paste,
they have the same instruction set and architecture.

File: cpu/cortex-m4_common/thread_arch.c (hunk 1 above)
Comment (gebart) [inline cpu/cortex-m4_common/thread_arch.c]: Maybe I should add a normal (non-Doxygen) comment hinting that the definition (implementation) is found inside the naked function at the end?

**Comment by gebart** [comment]
@haukepetersen well it will always be possible to just copy and paste the isr_svc out from inside the isr_pendsv when the need arises.

**Comment by jfischer-phytec-iot** [comment]
It works, tested on kw2x

File: cpu/cortex-m4_common/thread_arch.c (hunk 1 above)
Comment (haukepetersen) [inline cpu/cortex-m4_common/thread_arch.c]: Hm, I think its fine.

File: cpu/cortex-m4_common/thread_arch.c (hunk 2 above)
Comment (haukepetersen) [inline cpu/cortex-m4_common/thread_arch.c]: never mind.

**Comment by haukepetersen** [comment]
works also fine on the stm32f4discovery. ACK and go!

**Pull request merged**